lib/sudo-pam/src/converse.rs

```diff
@@ -273,3 +274,156 @@ pub(crate) extern "C" fn converse<C: Converser>(
+        }
+
+        fn handle_error(&self, msg: &str) -> PamResult<()> {
+            panic!("{msg}")
+        }
+
+        fn handle_info(&self, _msg: &str) -> PamResult<()> {
+            Ok(())
+        }
+    }
+
+    // essentially do the inverse of the "conversation function"
+    fn dummy_pam(msgs: &[PamMessage], talkie: &pam_conv) -> Vec<Option<String>> {
+        let pam_msgs = msgs
+            .iter()
+            .map(|PamMessage { msg, style, .. }| pam_message {
+                msg: unsafe { sudo_cutils::into_leaky_cstring(msg) },
+                msg_style: *style as i32,
+            })
+            .collect::<Vec<_>>();
+        let ptrs = pam_msgs
+            .iter()
+            .map(|x| x as *const pam_message)
+            .collect::<Vec<_>>();
+
+        let mut raw_response = std::ptr::null::<pam_response>() as *mut _;
+        let conv_err = converse::<String>(
+            ptrs.len() as i32,
+            ptrs.as_ptr() as *mut _,
+            &mut raw_response,
+            talkie.appdata_ptr,
+        );
+
+        // deallocate the leaky strings
+        for rec in pam_msgs {
+            unsafe { libc::free(rec.msg as *mut _) }
+        }
+        if conv_err != 0 {
+            return vec![];
+        }
+
+        let result = msgs
+            .iter()
+            .enumerate()
+            .map(|(i, _)| unsafe {
+                let ptr = raw_response.add(i);
+                if (*ptr).resp.is_null() {
+                    None
+                } else {
+                    // "The resp_retcode member of this struct is unused and should be set to zero."
+                    assert_eq!((*ptr).resp_retcode, 0);
+                    Some(sudo_cutils::string_from_ptr((*ptr).resp))
+                }
+            })
+            .collect();
+
+        unsafe { libc::free(raw_response as *mut _ as *mut _) };
+        result
+    }
+
+    fn msg(style: PamMessageStyle, msg: &str) -> PamMessage {
+        let msg = msg.to_string();
+        PamMessage {
+            style,
+            msg,
+            response: None,
+        }
+    }
+
+    // sanity check on the test cases; lib.rs is expected to manage the lifetime of the pointer
+    // inside the pam_conv object explicitly.
+
+    use std::marker::PhantomData;
+    struct PamConvBorrow<'a> {
+        pam_conv: pam_conv,
+        _marker: std::marker::PhantomData<&'a ()>,
+    }
+
+    impl<'a> PamConvBorrow<'a> {
+        fn new<C: Converser>(data: Pin<&'a mut ConverserData<C>>) -> PamConvBorrow<'a> {
+            PamConvBorrow {
+                pam_conv: unsafe { data.create_pam_conv() },
+                _marker: PhantomData,
+            }
+        }
+
+        fn borrow(&self) -> &pam_conv {
+            &self.pam_conv
+        }
+    }
+
+    #[test]
+    fn pam_gpt() {
+        let mut hello = ConverserData {
+            converser: "tux".to_string(),
+            panicked: false,
+        };
+        let cookie = PamConvBorrow::new(Pin::new(&mut hello));
```

Comment:
As this line demonstrated, the act of pinning a `ConverserData` doesn't actually have any effect (since it is `Unpin`).